building_sim_plugins/building_sim_common/include/building_sim_common/door_common.hpp

```diff
@@ -0,0 +1,244 @@
+      door_element, "type", door_type))
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      " -- Missing required parameters for [%s] plugin",
+      door_name.c_str());
+    return nullptr;
+  }
+
+  if ((left_door_joint_name == "empty_joint" &&
+      right_door_joint_name == "empty_joint") ||
+      (left_door_joint_name.empty() && right_door_joint_name.empty()))
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      " -- Both door joint names are missing for [%s] plugin, at least one"
+      " is required", door_name.c_str());
+    return nullptr;
+  }
+
+    // Default values for joint limts
+  double left_joint_lower_limit = -1.57;
+  double left_joint_upper_limit = 0.0;
+  double right_joint_lower_limit = 0.0;
+  double right_joint_upper_limit = 1.57;
+
+  auto extract_limits = [&](SdfPtrT& joint_sdf)
+  {
+    auto joint_sdf_clone = joint_sdf->Clone();
+    std::string joint_name;
+    get_sdf_attribute_required<std::string>(
+      joint_sdf_clone, "name", joint_name);
+    auto element = joint_sdf_clone;
+    get_element_required(joint_sdf_clone, "axis", element);
+    get_element_required(element, "limit", element);
+    if (joint_name == left_door_joint_name)
+    {
+      get_sdf_param_if_available<double>(element, "lower", left_joint_lower_limit);
+      get_sdf_param_if_available<double>(element, "upper", left_joint_upper_limit);
+      RCLCPP_INFO(node->get_logger(),
+        "Joint [%s] lower [%f] upper [%f]",
+        joint_name.c_str(),left_joint_lower_limit, left_joint_upper_limit);
+    }
+    else if (joint_name == right_door_joint_name)
+    {
+      get_sdf_param_if_available<double>(element, "lower", right_joint_lower_limit);
+      get_sdf_param_if_available<double>(element, "upper", right_joint_upper_limit);
+      RCLCPP_INFO(node->get_logger(),
+        "Joint [%s] lower [%f] upper [%f]",
+        joint_name.c_str(),left_joint_lower_limit, left_joint_upper_limit);
+    }
+  };
+
+  // Get the joint limits from parent sdf
+  auto parent = sdf->GetParent();
+  if (!parent)
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      "Unable to access parent sdf to retrieve joint limits");
+    return nullptr;
+  }
+
+  auto joint_element = parent->GetElement("joint");
+  if (!joint_element)
+  {
+    RCLCPP_ERROR(node->get_logger(),
+      "Parent sdf missing required joint element");
+    return nullptr;
+  }
+
+  extract_limits(joint_element);
+  // Find next joint element if present
+  joint_element = joint_element->GetNextElement("joint");
+  if (joint_element)
+    extract_limits(joint_element);
+
+  std::array<double,2> left_joint_limits ={
+    left_joint_lower_limit, left_joint_upper_limit};
+  std::array<double,2> right_joint_limits ={
+    right_joint_lower_limit, right_joint_upper_limit};
+
+  std::shared_ptr<DoorCommon> door_common(new DoorCommon(
+      door_name,
+      node,
+      params,
+      left_door_joint_name,
+      right_door_joint_name,
+      left_joint_limits,
+      right_joint_limits));
```

Comment:
You could simplify the DoorCommon constructor if it took a unordered_map<std::string, std::array<double,2>> that maps joint_name -> limits then you only need one argument instead of four. This map can be populated with the outputs of the refactored lambda function that returns pairs.
Also you won't need to check again in the door constructor for the validity of the joint, you can just iterate through the data structure.

Reply:
Modified such that DoorCommon constructor now accepts an `std::unordered_map<std::string, DoorCommon::DoorElement>`. This is directly used to initialize `_doors`.